include/rocksdb/utilities/replayer.h

```diff
@@ -0,0 +1,68 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <memory>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/status.h"
+#include "rocksdb/trace_record.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+struct ReplayOptions {
+  // Number of threads used for replaying. If 0 or 1, replay using
+  // single thread.
+  uint32_t num_threads;
+
+  // Enables fast forwarding a replay by increasing/reducing the delay between
+  // the ingested traces.
+  //   If > 0.0 and < 1.0, slow down the replay by this amount.
+  //   If 1.0, replay the operations at the same rate as in the trace stream.
+  //   If > 1, speed up the replay by this amount.
+  double fast_forward;
+
+  ReplayOptions() : num_threads(1), fast_forward(1.0) {}
+  ReplayOptions(uint32_t num_of_threads, double fast_forward_ratio)
+      : num_threads(num_of_threads), fast_forward(fast_forward_ratio) {}
+};
+
+// Replayer helps to replay the captured RocksDB query level operations.
+// The Replayer can either be created from DB::NewReplayer method, or be
+// instantiated via db_bench today, on using "replay" benchmark.
+class Replayer {
```

Comment:
I am not sure I understand this interface class.  I can see why/where Prepare and Replay are invoked publicly but I do not see any cases where Next and Execute are used outside of the Replay implementation.

Is there a need for those two methods publicly?  What is the use case?  Without them in the public API this PR could be greatly simplified as there would be many fewer public moving parts (TraceRecord and TraceType would no longer need to be public).

Reply:
We need 2 sets of APIs for replay. One is auto-replay, following timestamps thus controlled by number of threads and fast_forward speed. The other set is for manual replay, which basically does 2 things, one is decoding some string (acquired from TraceReader) to a TraceRecord (via Trace internally), the other is executing it (actually this is not necessary since user know the contents of the operation, and he/she can just call the corresponding DB APIs to execute them).